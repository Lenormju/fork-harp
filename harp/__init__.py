from ._logging import get_logger
from .factory import ProxyFactory

__all__ = [
    ProxyFactory,
    get_logger,
]
