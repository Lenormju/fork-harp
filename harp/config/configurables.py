from typing import Annotated, Generic, Self, Type, TypeVar

from pydantic import BaseModel, Field, field_serializer


class BaseConfigurable(BaseModel):
    @classmethod
    def from_dict(cls: Type[Self], data: dict) -> Self:
        return cls(**data)

    @classmethod
    def from_kwargs(cls: Type[Self], **kwargs) -> Self:
        return cls(**kwargs)


class Configurable(BaseConfigurable):
    class Config:
        extra = "forbid"


TConfigurable = TypeVar("TConfigurable", bound=Configurable)


class Stateful(BaseConfigurable, Generic[TConfigurable]):
    settings: Annotated[TConfigurable, Field(repr=False)]

    @classmethod
    def get_settings_type(cls) -> Type[Configurable]:
        return cls.model_fields["settings"].annotation

    @classmethod
    def from_settings_dict(cls: Type[Self], data: dict) -> Self:
        settings_type = cls.get_settings_type()
        return cls(settings=settings_type(**data))

    @classmethod
    def from_settings_kwargs(cls: Type[Self], **kwargs) -> Self:
        settings_type = cls.get_settings_type()
        return cls(settings=settings_type(**kwargs))

    @field_serializer("settings", when_used="json")
    @classmethod
    def __serialize_settings(cls, settings: TConfigurable):
        BaseType = type(settings).__mro__[type(settings).__mro__.index(Configurable) - 1]
        return BaseType.model_construct(
            **{k: v for k, v in settings.model_dump().items() if k in BaseType.model_fields}
        ).model_dump()
