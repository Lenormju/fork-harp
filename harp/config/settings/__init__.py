from .base import Settings, asdict, settings_dataclass
from .disabled import DisableableBaseSettings, DisabledSettings
from .lazy import Definition, Lazy

__all__ = [
    "Settings",
    "Definition",
    "DisableableBaseSettings",
    "DisabledSettings",
    "Lazy",
    "asdict",
    "settings_dataclass",
]
