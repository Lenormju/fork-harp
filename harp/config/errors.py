from harp.errors import BaseError


class ConfigurationError(BaseError):
    pass


class ConfigurationValueError(ConfigurationError, ValueError):
    pass


class ConfigurationRuntimeError(ConfigurationError, RuntimeError):
    pass


class ConfigurationRemovedSettingError(ConfigurationError):
    pass
