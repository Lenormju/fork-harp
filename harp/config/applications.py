from typing import Type

from rodi import Container
from whistle import IAsyncEventDispatcher

from .events import EVENT_FACTORY_BIND, EVENT_FACTORY_BOUND, EVENT_FACTORY_BUILD, EVENT_FACTORY_DISPOSE
from .settings.base import asdict
from .utils import get_application_type, resolve_application_name


class Application:
    class Settings(dict):
        pass

    class Lifecycle:
        on_bind = None
        """
        Placeholder for factory bind event, happening before the container is built. If set, it will be attached to the
        factory dispatcher automatically.

        """

        on_bound = None
        """
        Placeholder for factory bound event, happening after the container is built. If set, it will be attached to the
        factory dispatcher automatically.
        """

        on_build = None
        """
        Placeholder for factory build event, happening after the kernel is built. If set, it will be attached to the
        factory dispatcher automatically.
        """

        on_dispose = None
        """
        Placeholder for factory dispose event, happening after the kernel is disposed. If set, it will be attached to the
        factory dispatcher automatically, in reverse order of appearance (first loaded application will be disposed last).
        """


class ApplicationsRegistry:
    def __init__(self):
        self._applications = {}

    def __contains__(self, name):
        return name in self._applications

    def __getitem__(self, name):
        return self._applications[name]

    def __iter__(self):
        yield from self._applications

    def __len__(self):
        return len(self._applications)

    def get_full_name(self, name):
        mod_name = self[name].__module__
        if mod_name.endswith(".__app__"):
            return mod_name[:-8]
        return mod_name + "." + self[name].__qualname__

    def add(self, *names):
        for name in names:
            full_name = resolve_application_name(name)
            short_name = full_name.split(".")[-1]

            if short_name not in self._applications:
                self._applications[short_name] = get_application_type(full_name)
            elif self._applications[short_name] != get_application_type(full_name):
                raise ValueError(
                    f"Application {short_name} already registered with a different type ({self._applications[short_name].__module__}.{self._applications[short_name].__qualname__})."
                )

    def remove(self, *names):
        for name in names:
            full_name = resolve_application_name(name)
            short_name = full_name.split(".")[-1]

            if short_name in self._applications:
                del self._applications[short_name]

    def items(self):
        return self._applications.items()

    def keys(self):
        return self._applications.keys()

    def values(self):
        return self._applications.values()

    def lifecycle_for(self, name) -> Type[Application.Lifecycle]:
        return getattr(self._applications[name], "Lifecycle") or Application.Lifecycle

    def settings_for(self, name) -> Type[Application.Settings]:
        return getattr(self._applications[name], "Settings") or dict

    def defaults(self):
        return {name: asdict(self.settings_for(name)()) for name in self._applications}

    def register_events(self, dispatcher: IAsyncEventDispatcher):
        for name, AppType in self.items():
            lifecycle = getattr(AppType, "Lifecycle", None) or Application.Lifecycle

            if on_bind := getattr(lifecycle, "on_bind", None):
                dispatcher.add_listener(EVENT_FACTORY_BIND, on_bind)

            if on_bound := getattr(lifecycle, "on_bound", None):
                dispatcher.add_listener(EVENT_FACTORY_BOUND, on_bound)

            if on_build := getattr(lifecycle, "on_build", None):
                dispatcher.add_listener(EVENT_FACTORY_BUILD, on_build)

        for name, AppType in reversed(self.items()):
            lifecycle = getattr(AppType, "Lifecycle", None) or Application.Lifecycle

            if on_dispose := getattr(lifecycle, "on_dispose", None):
                dispatcher.add_listener(EVENT_FACTORY_DISPOSE, on_dispose)

    def register_services(self, container: Container, config: dict):
        for name, AppType in self.items():
            settings_type = self.settings_for(name)
            local_config = config.get(name, None)
            if settings_type is not dict and local_config:
                container.add_instance(local_config, settings_type)

    def aslist(self):
        return [self.get_full_name(name) for name in self]
