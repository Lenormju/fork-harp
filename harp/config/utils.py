import importlib.util
from typing import TYPE_CHECKING, Type

from .defaults import DEFAULT_NAMESPACES

if TYPE_CHECKING:
    from .applications import Application


def get_application_class_name(name):
    return "".join(map(lambda x: x.title().replace("_", ""), name.rsplit(".", 1)[-1:])) + "Application"


def resolve_application_name(spec):
    if "." not in spec:
        for namespace in DEFAULT_NAMESPACES:
            _candidate = ".".join((namespace, spec))
            if importlib.util.find_spec(_candidate):
                return _candidate
    if importlib.util.find_spec(spec):
        return spec

    raise ModuleNotFoundError(f"No application named {spec}.")


#: Cache for application types
_application_types = {}


def get_application_type(name: str) -> Type["Application"]:
    """
    Returns the application class for the given application name.

    :param name:
    :return:
    """
    if name not in _application_types:
        application_spec = importlib.util.find_spec(name)
        if not application_spec:
            raise ValueError(f'Unable to find application "{name}".')

        try:
            application_module = __import__(".".join((application_spec.name, "__app__")), fromlist=["*"])
        except ModuleNotFoundError as exc:
            raise ModuleNotFoundError(
                f'A python package for application "{name}" was found but it is not a valid harp application. '
                'Did you forget to add an "__app__.py"?'
            ) from exc

        application_class_name = get_application_class_name(name)
        if not hasattr(application_module, application_class_name):
            raise AttributeError(f'Application module for {name} does not contain a "{application_class_name}" class.')

        _application_types[application_spec.name] = getattr(application_module, application_class_name)
        _application_types[application_spec.name].name = application_spec.name

    if name not in _application_types:
        raise RuntimeError(f'Unable to load application "{name}", application class definition not found.')

    return _application_types[name]
