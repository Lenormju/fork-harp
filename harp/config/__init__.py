"""
The Config (:mod:`harp.config`) module provides tools to configure *Applications*.

*Applications* are standard python packages that contains specific definitions to be able to register services and
configure themselves. Every feature of Harp is built as an application, and you will want to write applications if you
want to extend harp's functionality for specific needs.


Contents
--------

"""

__title__ = "Config"

from .applications import Application, ApplicationsRegistry
from .builders import ConfigurationBuilder, System, SystemBuilder
from .events import (
    EVENT_FACTORY_BIND,
    EVENT_FACTORY_BOUND,
    EVENT_FACTORY_BUILD,
    EVENT_FACTORY_DISPOSE,
    FactoryBindEvent,
    FactoryBoundEvent,
    FactoryBuildEvent,
    FactoryDisposeEvent,
)
from .settings import (
    BaseSetting,
    Definition,
    DisableableBaseSettings,
    DisabledSettings,
    FromFileSetting,
    Lazy,
    asdict,
    settings_dataclass,
)
from .utils import get_application_type

__all__ = [
    "Application",
    "ApplicationsRegistry",
    "BaseSetting",
    "ConfigurationBuilder",
    "Definition",
    "DisableableBaseSettings",
    "DisabledSettings",
    "EVENT_FACTORY_BIND",
    "EVENT_FACTORY_BOUND",
    "EVENT_FACTORY_BUILD",
    "EVENT_FACTORY_DISPOSE",
    "FactoryBindEvent",
    "FactoryBoundEvent",
    "FactoryBuildEvent",
    "FactoryDisposeEvent",
    "FromFileSetting",
    "Lazy",
    "System",
    "SystemBuilder",
    "asdict",
    "get_application_type",
    "settings_dataclass",
]
