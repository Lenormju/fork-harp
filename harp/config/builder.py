import os

from config.common import ConfigurationBuilder as BaseConfigurationBuilder
from config.common import ConfigurationSource


class ConfigurationBuilder(BaseConfigurationBuilder):
    def __init__(self, *sources: ConfigurationSource) -> None:
        super().__init__(*sources)

    def add_examples(self, examples):
        from config.yaml import YAMLFile

        from harp.config import examples as examples_package

        _examples_dirname = os.path.dirname(examples_package.__file__)
        for example in examples or ():
            _example_file = os.path.join(_examples_dirname, f"{example}.yml")
            self.add_source(YAMLFile(_example_file))

    def add_files(self, files):
        for file in files or ():
            _, ext = os.path.splitext(file)
            if ext in (".yaml", ".yml"):
                from config.yaml import YAMLFile

                self.add_source(YAMLFile(file))
            elif ext in (".json",):
                from config.json import JSONFile

                self.add_source(JSONFile(file))
            elif ext in (".ini", ".conf"):
                from config.ini import INIFile

                self.add_source(INIFile(file))
            elif ext in (".toml",):
                from config.toml import TOMLFile

                self.add_source(TOMLFile(file))
            else:
                raise ValueError(f"Unknown file extension: {ext}")

    def add_values(self, values: dict):
        for k, v in values.items():
            self.add_value(k, v)
