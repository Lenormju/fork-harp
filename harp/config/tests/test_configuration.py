import orjson

from harp.config import Application, ApplicationsRegistry, ConfigurationBuilder, asdict
from harp.utils.identifiers import is_valid_dotted_identifier


def test_is_valid_extension_name():
    # empty
    assert not is_valid_dotted_identifier("")

    # usual
    assert is_valid_dotted_identifier("foo.bar")
    assert is_valid_dotted_identifier("toto")

    # empty part
    assert not is_valid_dotted_identifier(".foo")
    assert not is_valid_dotted_identifier("foo..bar")
    assert not is_valid_dotted_identifier("bar.")

    # dashed/underscored
    assert is_valid_dotted_identifier("o_o")
    assert not is_valid_dotted_identifier("o-o")
    assert is_valid_dotted_identifier("_mazette")
    assert is_valid_dotted_identifier("raise_")


class ApplicationsRegistryMock(ApplicationsRegistry):
    def add_mock(self, name, impl):
        self._applications[name] = impl


def test_add_application(snapshot):
    builder = ConfigurationBuilder(use_default_applications=False)
    builder.applications.add("storage")

    assert len(builder.applications) == 1
    assert builder.applications["storage"].__module__ == "harp_apps.storage.__app__"

    settings = builder.build()

    serialized = orjson.dumps(asdict(settings))
    assert serialized == snapshot

    new_builder = ConfigurationBuilder.from_bytes(serialized, ApplicationsRegistryType=ApplicationsRegistryMock)
    assert asdict(new_builder.build()) == asdict(settings)

    new_builder.applications.add_mock("foo.bar", type("MockedApplication", (Application,), {"name": "foo.bar"}))

    new_settings = new_builder.build()
    assert asdict(new_settings) != asdict(settings)

    assert asdict(new_settings) == snapshot
    assert orjson.dumps(asdict(new_settings)) == snapshot
