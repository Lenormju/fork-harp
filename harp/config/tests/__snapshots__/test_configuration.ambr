# serializer version: 1
# name: test_add_application
  b'{"applications":["harp_apps.storage"],"storage":{"url":"sqlite+aiosqlite:///:memory:?cache=shared","migrate":true,"blobs":{"type":"sql"}}}'
# ---
# name: test_add_application.1
  dict({
    'applications': list([
      'harp_apps.storage',
      'harp.config.tests.test_configuration.MockedApplication',
    ]),
    'storage': dict({
      'blobs': dict({
        'type': 'sql',
      }),
      'migrate': True,
      'url': 'sqlite+aiosqlite:///:memory:?cache=shared',
    }),
  })
# ---
# name: test_add_application.2
  b'{"applications":["harp_apps.storage","harp.config.tests.test_configuration.MockedApplication"],"storage":{"url":"sqlite+aiosqlite:///:memory:?cache=shared","migrate":true,"blobs":{"type":"sql"}}}'
# ---
